rclc/src/rclc/executor.c

```diff
@@ -1067,6 +1501,127 @@ _rclc_execute(rclc_executor_handle_t * handle)
+            NULL != goal_handle)
+          {
+            goal_handle->available_goal_response = false;
+            handle->action_client->goal_callback(
+              goal_handle, goal_handle->goal_accepted,
+              handle->callback_context);
+            if (!goal_handle->goal_accepted ||
+              RCL_RET_OK != rclc_action_send_result_request(goal_handle))
+            {
+              rclc_action_put_goal_handle(handle->action_client, goal_handle);
+            } else {
+              goal_handle->status = GOAL_STATE_ACCEPTED;
+            }
+          }
+        }
+        if (handle->action_client->feedback_available) {
+          rclc_action_goal_handle_t * goal_handle;
+          while (goal_handle = rclc_action_find_first_handle_with_feedback(handle->action_client),
+            NULL != goal_handle)
+          {
+            goal_handle->available_feedback = false;
+            handle->action_client->feedback_callback(
+              goal_handle,
+              handle->action_client->ros_feedback,
+              handle->callback_context);
+          }
+        }
+        if (handle->action_client->cancel_response_available) {
+          rclc_action_goal_handle_t * goal_handle;
+          while (goal_handle =
+            rclc_action_find_first_handle_with_cancel_response(handle->action_client),
+            NULL != goal_handle)
+          {
+            goal_handle->available_cancel_response = false;
+            if (handle->action_client->cancel_callback != NULL) {
+              handle->action_client->cancel_callback(
+                goal_handle,
+                goal_handle->goal_cancelled,
+                handle->callback_context);
+            }
+          }
+        }
+        if (handle->action_client->result_response_available) {
+          rclc_action_goal_handle_t * goal_handle;
+          while (goal_handle =
+            rclc_action_find_first_handle_with_result_response(handle->action_client),
+            NULL != goal_handle)
+          {
+            goal_handle->available_result_response = false;
+            handle->action_client->result_callback(
+              goal_handle,
+              handle->action_client->ros_result_response,
+              handle->callback_context);
+            rclc_action_put_goal_handle(handle->action_client, goal_handle);
+          }
+        }
+        break;
+
+      case ACTION_SERVER:
+        if (handle->action_server->goal_ended) {
+          rclc_action_goal_handle_t * goal_handle;
+          while (goal_handle =
+            rclc_action_find_first_terminated_handle(handle->action_server),
+            NULL != goal_handle)
+          {
+            rclc_action_put_goal_handle(goal_handle->action_server, goal_handle);
+          }
+          handle->action_server->goal_ended = false;
+        }
+        if (handle->action_server->goal_request_available) {
+          rclc_action_goal_handle_t * goal_handle;
+          while (goal_handle =
```

Comment:
this while loop worries me! 

It is not obvious, that the `handle->status` is modified, so that a handle with `GOAL_STATE_UNKNOWN` is not returned multiple times resulting in an endless while-loop.
 
Please add a description, e.g. as comment, that indeed this while loop always terminates. Please also check termination condition for all such while loops in for action_server and action_client code.

Reply:
The goal is either removed from `used_goal_handles` here:
https://github.com/ros2/rclc/blob/30ba172cf03d3da3fd3d243efc1df732c7dbb401/rclc/src/rclc/executor.c#L1596-L1599

or its `handle->status` gets modified here:
https://github.com/ros2/rclc/blob/30ba172cf03d3da3fd3d243efc1df732c7dbb401/rclc/src/rclc/executor.c#L1600-L1603

We should free the goal on the default switch case.

Reply:
Include default case as goal rejected to avoid infinite loop: https://github.com/ros2/rclc/pull/170/commits/3b935542f05d350b5f4698820759a373f345393

I will also check termination conditions for all loops carefully, adding comments in each exit case.

Reply:
Added comments on loops: 04309951f7c2ac129c0f7d8fba0b48f6f74b1886

Reply:
@Acuadros95 I my view, the documentation in the comments is not sufficient to verify that the while loop always terminates. Could you please elaborate more in detail like this:
pre-condition: 
status = GOAL_STATE_XXX
post-condition:
status = xxx
 for each possible states in the particular while loops?

You need to make sure that in each case, that the number of elementes in the loop is reduced by one in each iteration. And indeed, the status is changed in such a way, that you will not end up in an endless loop. (e.g. the status is never updated then the goal will remain forever in the list and this would end up in an endless loop.

Reply:
@JanStaschulat Update: 0c2e552b6489a3f24e32c20de28790b6fbd84ae6

I think we covered all the cases for each loop, as we encountered this kind of bug on the testing phase.